### ISSUE TITLE
Question about Loading Quantized Models and Extracting Scaling Factors

### ISSUE THREAD

**Issue by RealJustinNi** [Xilinx/brevitas#1176]
Hi! I'm exploring the quantized model generated via brevitas/src/brevitas_examples/imagenet_classification/ptq/ptq_evaluate.py. Thank you for your excellent work, the quantized model achieved good accuracy during validation.

As part of further analysis, I’d like to:
1. Re-run evaluation on a full test set using a standalone script (separate from the original codebase), and
2. Observe the scaling factors and the distributions of intermediate activations and layer weights to better understand the quantized model's behavior.

Could you provide guidance on how to load the saved quantized model (e.g., .torch_qcdq files) with scaling factors and perform inference?

Any guidance  would be greatly appreciated, thanks!

**Comment by Giuseppe5** [comment]
Hey, thanks for the nice words!

> Re-run evaluation on a full test set using a standalone script (separate from the original codebase)

Currently we are slowly adding this support to the various entry-points (it is possible to do this with SDXL and LLM, not with the imagenet PTQ one yet unfortunately). It is not overly complicate, however it might have side effects based on what PTQ algorithms you have used.
In general, it should be as simple as:
- loading the floating point model
- re-apply the same quantization
- load the state dict

However some algorithms modify the graph structure/state dict, so you need to re-apply them before quantization/reloading of the state dict. It might be difficult to read, but if you see our stable_diffusion/main.py entrypoint and follow the `load_checkpoint` flag, you can see what I mean.

I would say, we can help with adapting the current script to load a checkpoint, but for more specific solutions, I'd recommend creating your own custom script.

> torch_qcdq

In the answer above, I focused mostly on state_dict since we use torch_qcdq mostly as a way to represent the computational graph, rather than a way to perform inference. However, we should have some tests that you can check to see how to load/run a torch_qcdq model and use that for inference.

> Observe the scaling factors and the distributions of intermediate activations and layer weights to better understand the quantized model's behavior.


We don't have anything ready for this unfortunately, and my best advice would be to use torch forward hooks to accumulate statistics. To make the matter a bit easier, you can add the quantization nodes, then disabling the quantization (we have a utility class in brevitas/graph/calibrate.py), and then attach your hooks to the forward of the proxy, so that you can see the statistics in the point of the graph that are relevant for the quantization process.



Let me know if this helped you!

**Comment by RealJustinNi** [comment]
Thank you for your reply!
In fact, for torchvision's PTQ models, I later analyzed the exported ONNX model using Netron and observed its nodes, which allowed me to extract all the weight parameters and scaling factors. During inference, I can also obtain the activation distribution by observing node outputs in a similar way. 
However, I'm still exploring transformer-like APIs. For example, when I directly quantize BERT and export it to ONNX, I might encounter errors indicating that some operators are invalid or unsupported. I still need some time to prepare detailed information and a reproducible script for this issue, but if you're interested, I will provide it to you promptly in the future.
 Lastly, thank you again for your contributions to this project and for your help with my questions! 👍

**Comment by Giuseppe5** [comment]
Happy to help!

I am going to mark this as solved, but feel free to re-open in case I missed something.
